### PR TITLE
Stepper controller node

### PR DESCRIPTION
**Pull request by Buzzzsaw**
Working stepper controller node in the Catkin workspace.
Still some testing to do with the SPI controller, and some improvements are still to be made to it.

### CODE REVIEW

File: stepper/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  
+## Stepper Controller node
+
+Start node:
+```
+rosrun stepper_controller stepper_controller_node
+```
+
+This node listens to publishers on:
+ * /stepper_controller_run : issue a command to an LED (on, off, blink)
```
Comment (mgaudreault) [inline stepper/README.md]: Stepper LED ?

File: stepper/stepper_controller/src/loltest.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#! /usr/bin/env python
```
Comment (mgaudreault) [inline stepper/stepper_controller/src/loltest.py]: Please change filename to something relevant or remove the file

**Review by mgaudreault — COMMENTED** [review]
Minor changes to do

File: stepper/README.md (hunk 1 above)
Comment (Buzzzsaw) [reply stepper/README.md]: Whoops! Copy paste error, good catch.

File: stepper/stepper_controller/src/loltest.py (hunk 2 above)
Comment (Buzzzsaw) [reply stepper/stepper_controller/src/loltest.py]: Will be removed since not useful anymore.

File: stepper/README.md (hunk 1 above)
Comment (Buzzzsaw) [reply stepper/README.md]: Improved README au passage with the latest changes.

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+        StepSec2 = StepTick2 / self.STEP_ACC_RATIO
+        
+        return StepSec2
+
+    # Convert an acceleration from step/s^2 to step/tick^2.
+    def accToStepTick(self, StepSec2):
+        # Controller expects an acceleration in step/tick^2 formatted in unsigned fixed point 0.40.
+        # Equation: StepTick2 = (StepSec2 * TICK^2)/2^-40
+        # Equivalent to: StepTick2 * 0.068719
+        StepTick2 = int(StepSec2 * self.STEP_ACC_RATIO)
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: int() is the equivalent for floor(). If precision is important here we should probably use math.round()

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+        StepSec = StepTick / self.STEP_SPEED_RATIO
+        
+        return StepSec
+
+    # Convert a speed from step/s to step/tick.
+    def speedToStepTick(self, StepSec):
+        # Controller expects speed in step/tick formatted in unsigned fixed point 0.28.
+        # Equation: StepTick = (StepSec * TICK)/(2^-28)
+        # Equivalent to: StepSec * 67.108864
+        StepTick = int(StepSec * self.STEP_SPEED_RATIO)
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: int() is the equivalent for floor(). If precision is important here we should probably use math.round()

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+        # Controller expects an acceleration in step/tick^2 formatted in unsigned fixed point 0.40.
+        # Equation: StepTick2 = (StepSec2 * TICK^2)/2^-40
+        # Equivalent to: StepTick2 * 0.068719
+        StepTick2 = int(StepSec2 * self.STEP_ACC_RATIO)
+        
+        # Limit speed to a 12 bits number
+        if StepTick2 > 0x00000FFF:
+            return 0x00000FFF
+        
+        return StepTick2
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: return min(StepTick2, 0x00000FFF)

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 6]
```diff
@@ -0,0 +1,563 @@
...
+        # Controller expects speed in step/tick formatted in unsigned fixed point 0.28.
+        # Equation: StepTick = (StepSec * TICK)/(2^-28)
+        # Equivalent to: StepSec * 67.108864
+        StepTick = int(StepSec * self.STEP_SPEED_RATIO)
+        
+        # Limit speed to a 20 bits number
+        if StepTick > 0x000FFFFF:
+            return 0x000FFFFF
+        
+        return StepTick
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: return min(StepTick, 0x000FFFFF)

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 7]
```diff
@@ -0,0 +1,563 @@
...
+        # Controller expects speed in step/tick formatted in unsigned fixed point 0.18.
+        # Equation: StepTick = (StepSec * TICK)/(2^-18)
+        # Equivalent to: StepSec * 0.065536
+        StepTick = StepSec * self.MAXMIN_STEP_SPEED_RATIO
+        
+        # Limit speed to a 20 bits number
+        if StepTick > 0x000FFFFF:
+            return 0x000FFFFF
+        
+        return StepTick
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: return min(StepTick, 0x000FFFFF)

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 8]
```diff
@@ -0,0 +1,563 @@
...
+    # SetParam command
+    # Set the given register to the given value.
+    def setParam(self, Param, Value):
+        # Only accept Param as a 5 bits number.
+        if Param > 0x1F:
+            return
+
+        # Limit Value to a 24 bits number.
+        if Value > 0x00FFFFFF: 
+            Value = 0x00FFFFFF
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: Value = min(Value, 0x00FFFFFF)

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 9]
```diff
@@ -0,0 +1,563 @@
...
+    # (full, half, quarter, etc.).
+    def move(self, Direction, Steps):
+        if Direction != self.DIR_CLOCKWISE and Direction != self.DIR_COUNTER_CLOCKWISE:
+            return
+        if Steps <= 0:
+            return
+                
+        # Limit steps to a 22 bits number
+        if Steps > 0x003FFFFF:
+            Steps = 0x003FFFFF
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: Steps = min(Steps, 0x003FFFFF)

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 10]
```diff
@@ -0,0 +1,563 @@
...
+        self.sendCmd3(self.CMD_MOVE | Direction, int(Steps))
+
+    # GoTo command
+    # Move the stepper to the given absolute position in agreement with the selected step mode 
+    # (full, half, quarter, etc.).
+    # This function uses the shortest path possible to get to that position.
+    def goTo(self, Position):
+        # Limit posiition to a 22 bits number
+        if Position > 0x003FFFFF:
+            Position = 0x003FFFFF
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: Position = min(Position, 0x003FFFFF)

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 11]
```diff
@@ -0,0 +1,563 @@
...
+    # GoToDir command
+    # Same ad GoTo, but with a forced rotation direction, which depending on that direction might 
+    # not result in the shortest path.
+    def goToDir(self, Direction, Position):
+        if Direction != self.DIR_CLOCKWISE or Direction != self.DIR_COUNTER_CLOCKWISE:
+            return
+
+        # Limit posiition to a 22 bits number
+        if Position > 0x003FFFFF:
+            Position = 0x003FFFFF
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: Position = min(Position, 0x003FFFFF)

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 12]
```diff
@@ -0,0 +1,563 @@
...
+    # Set a new minimum speed for the stepper, in step/s.
+    # The Optimization parameter set the low-speed optimization flag. If set to True, low-speed
+    # optimization will be activated and the minimum speed value will be ignored by the controller.
+    def setMinSpeed(self, MinSpeed, Optimized):
+        if MinSpeed < 0:
+            return
+        
+        # Limit value to 12 bits.
+        if MinSpeed > 0x0FFF:
+            MinSpeed = 0x0FFF
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: MinSpeed = min(MinSpeed, 0x0FFF)

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 13]
```diff
@@ -0,0 +1,563 @@
...
+
+    # SetMaxSpeed command
+    # Set a new maximum speed for the stepper, in step/s.
+    def setMaxSpeed(self, MaxSpeed):
+        if MaxSpeed < 0:
+            return
+        
+        # Limit value to 10 bits.
+        if (MaxSpeed > 0x03FF):
+            MaxSpeed = 0x03FF
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: MaxSpeed = min(MaxSpeed, 0x03FF)

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 14]
```diff
@@ -0,0 +1,563 @@
...
+        if MinSpeed < 0:
+            return
+        
+        # Limit value to 12 bits.
+        if MinSpeed > 0x0FFF:
+            MinSpeed = 0x0FFF
+        
+        # Add the low speed optimization bit if specified.
+        if Optimized == True:
+            MinSpeed += 0x1000
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py:507]: Your way is probably cleaner, but here's a neat python golf trick:
`MinSpeed += (0, 0x1000)[Optimized]`

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 15]
```diff
@@ -0,0 +1,563 @@
...
+    # Set a new threshold speed for the stepper, in step/s.
+    # When the actual speed exceeds this value, the step mode is automatically switched to 
+    # full-step two-phase on.
+    def setThresholdSpeed(self, ThresholdSpeed):
+        if ThresholdSpeed < 0:
+            return
+        
+        # Limit value to 10 bits.
+        if ThresholdSpeed > 0x03FF:
+            ThresholdSpeed = 0x03FF
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: ThresholdSpeed = min(ThresholdSpeed, 0x03FF)

File: stepper/stepper_controller/include/L6470_pkg/L6470_lib.py [hunk 16]
```diff
@@ -0,0 +1,563 @@
...
+                
+        self.setParam(self.REG_STEP_MODE, SyncEn | SyncSel | StepSel)
+
+    # SetAlarms command
+    # Configure which alarms are active on the controller.
+    # See the ALARM values for options.
+    def setAlarms(self, Alarms):
+        # Limit alarms to 8 bits
+        if Alarms > 0xFF:
+            Alarms = 0xFF
```
Comment (carapas) [inline stepper/stepper_controller/include/L6470_pkg/L6470_lib.py]: Alarms = min(Alarms, 0xFF)

File: stepper/stepper_controller/src/stepper_controller_node.py [hunk 17]
```diff
@@ -0,0 +1,77 @@
+#! /usr/bin/env python
+
+# ROS packages.
+import rospy
+
+# Stepper generated messages.
+from stepper_controller.msg import Run
+from stepper_controller.msg import Move
+from stepper_controller.msg import GoTo
+from stepper_controller.msg import Stop
```
Comment (carapas) [inline stepper/stepper_controller/src/stepper_controller_node.py]: Importing theses messages in node.js might be a pain. Maybe using std_msgs.String and using JSON would be a more standardised way of communicating?

**Review by carapas — CHANGES_REQUESTED** [review]
Really great!
I would like to discuss about the JSON part of my review before pushing.

**Comment by hugbed** [comment]
Node should subscribe to topic : /motor/inter_eye/command

**Comment by mgaudreault** [comment]
Fixed hugbed request, but we need more error handling I think, might do that tomorrow.

**Pull request merged**